RFC: Sync manager

mhammond opened the pull request:
As promised, I wrote up an initial RFC for the sync-manager a few of us have been discussing. It's not complete, but there should be enough there to get initial feedback.

[Rendered](https://github.com/mozilla/application-services/blob/sync-manager-rfc/docs/rfcs/sync-manager.md)

rfk commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+
+    // Perform the sync of the engine. When asked to sync, the manager will
+    // create a GlobalEngineState specific to the engine and pass it in.
+    fn sync(state: GlobalEngineState) -> ???
```
For Firefox iOS, they current have code to pass in a (tokenserver_uri, oauth_token) pair to the logins `sync` method.  It seems reasonable to expect them to update that code to instead pass in a synthesized `GlobalEngineState` object created out of the state of their existing sync infrastructure.

For Lockbox, will we require them to create and drive a new `SyncManager` object of some sort?  Or will we maintain a higher-level API that matches the current behaviour, taking an oauth token and doing the sync-manager logic internally?  There's a risk here that we'll end up splintering the consumer experience for Lockbox on Android (where they can use the a-c `SyncManager` plumbing) and Lockbox on iOS (where they may not have as many nice abstractions).

mhammond replied:
> How does this `sync` method relate to the `sync` method on, say, the logins store that is already called by lockbox? Is it a replacement, or is it on a different object, or something else?

Both? :) I think this will be a replacement *and* on a different object. ie, lockbox will still have just 1 sync function to call, but it may obtain that function in a slightly different way. We can probably make this invisible to them though. I also hope that this will be the same on iOS.

I suspect that the general sync process will still take the same inputs as it does now, it's just that the responsibilities for consuming those inputs will shift a little. I'll try and be more specific about this in the next version.

thomcc commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+# Sync manager
+
+We've identified the need for a "sync manager" (although are yet to identify a good name for it!) This manager will be responsible for managing "global sync state" and coordinating each engine. For the purposes of this, we are including the "clients" engine in this global state.
```
Interesting that the clients engine is being considered, although I guess this is the only sane place for it.

thomcc commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+# Sync manager
+
```
nit: Probably worth making lines shorter so we can respond to individual points. 80 or 100 line columns are probably fine. There are good plugins for this (just hit alt+q with https://marketplace.visualstudio.com/items?itemName=stkb.rewrap or https://packagecontrol.io/packages/Wrap%20Plus).

mhammond replied:
yeah, I never know whether .md should be explicitly wrapped or not. I'll re-wrap it.

thomcc commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+
+* Manage interaction with the token server to obtain a token. As part of this, the manager needs to detect when an engine fails to use the token, generally due to a node reassignment mid sync. In this scenario, the manager will abort any pending syncs and reset itself, starting a new sync flow. Note also that this means the containing application will generally not need to see authentication erro ...
+
+* Manage the "clients" collection - we probably can't ignore this any longer, especially for bookmarks (as desktop will send a wipe command on bookmark restore). This involves both communicating with the engine regarding commands targetting it, and accepting commands to be send to other devices. Note however that it's not yet clear where these commands will originate from - eg, consider a bookmar ...
```
Ugh yeah, forgot about the fact that bookmarks is likely to need it for correctness.

linacambridge replied:
We can maybe punt on this for the Fenix beta (you'll just get all your old bookmarks back 🙃), but I think we'll need to handle client commands somehow before shipping. We talked about moving them over to push, but, given the need to support existing clients that restore, I think we'll need to have _something_ that handles them.

mhammond replied:
> We talked about moving them over to push

TBH I'm not convinced push is a good replacement for engine commands. OTOH, i suspect wipes are the only commands with value, so an alternative for that might be possible (but one doesn't spring to mind)

thomcc commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+
+* Manage the "clients" collection - we probably can't ignore this any longer, especially for bookmarks (as desktop will send a wipe command on bookmark restore). This involves both communicating with the engine regarding commands targetting it, and accepting commands to be send to other devices. Note however that it's not yet clear where these commands will originate from - eg, consider a bookmar ...
+
+* Perform, or coordinate, the actual sync of the engines - from the containing app's POV, there's a single "sync now" entry-point. Exactly how the manager binds to engines is TBD. For example, a possible implementation is that a callback is registered with the sync manager. It's also likely that we will expose a way to sync a specific engine rather than "sync all".
```
> For example, a possible implementation is that a callback is registered with the sync manager.

I'm unsure how this can be done safely. To be safe the callback would need to be `+ 'static`, which severely limits what we can do. (And that's assuming we're megazorded, it would be extremely unsafe outside of megazording)

I think realistically it will probably have to return a list of engines (e.g. probably strings or enums) over the FFI.

linacambridge replied:
One thing to consider is shipping the monolith first: for MVP, require Megazording, bundle all engines, and sync all, not a specific engine. This means Lockbox and other consumers will temporarily pull in code for bookmarks and history, even if they don't use them—though we could maybe disable storage for those engines, and our Sync adapter could detect that and skip syncing the engine?

Once we have the Sync manager in place, we could go back, see where the boundaries are, and figure out how to split them up. I think I'm a bit wary of trying to figure out how to modularize this now, especially when the FFI is involved, when we haven't built it.

mhammond replied:
> One thing to consider is shipping the monolith first: for MVP, require Megazording, bundle all engines, and sync all, not a specific engine.

I agree (and should have been more explicit about this). In the first cut, I expect that if the rust sync manager is used, all engines will be rust implemented - this will be true for Fenix and Lockbox. iOS (and later, desktop) is a special case where the manager is likely to not be rust implemented, but that seems relatively easy too - it can just make regular ffi calls into the rust engines.

mhammond replied:
I removed all references to callbacks and added some extra comments around this.

thomcc commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+
+* Perform, or coordinate, the actual sync of the engines - from the containing app's POV, there's a single "sync now" entry-point. Exactly how the manager binds to engines is TBD. For example, a possible implementation is that a callback is registered with the sync manager. It's also likely that we will expose a way to sync a specific engine rather than "sync all".
+
+* Manage the fact that there may not be a 1:1 relationship between collections and sync engines. For example, it may be true in the future that a single engine uses multiple collections. A possible example is when we come to syncing "containers" - the history engine may then use 2 collections as part of a sync.
```
History and bookmarks are a lot more closely related. Even in my generic record sync plan, different engines in the same backing store can basically be treated as completely distinct.

mhammond replied:
> IIUC we also kind of consider "forms" and "addresses"

I guess you mean "creditcards" and "addresses" - that's an interesting discussion which touches on Thom's "generic store" proposal, but the tl;dr of that is that the implementation of those engines was an exercise in code reuse rather than due to the stores being closely related logically. Stuff like reconciliation is quite separate.

For the purposes of what I wrote in the document above, I'm thinking of a data type which requires 2 collections to have a consistent single local store and is something that would desire cross-collection transaction support on the server.

TBH, I'm still torn as to whether we should support that initially given there's no actual use-case.  On one hand  I suspect containers are going to keep coming up, but on the other hand, it might be enough to have history and containers managed by separate engines so long as we can ensure the order of syncing them.

thomcc commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+
+* It is not a sync scheduler. The containing app will be responsible for knowing when to sync (but as above, the scheduler does know *what* to sync.
+
+(anything else?)
```
Hopefully there's quite a bit more it won't do!

But I'm not sure anything else is worth calling out.

Maybe something about telemetry? Or will it do that?

linacambridge replied:
* It could manage telemetry aggregation, but not submission. (Up to the app to figure out how to handle the telemetry it collects).
* It doesn't handle FxA authentication, or manage signed in user state. (But it does talk to the token server!)
* It doesn't handle collection validation. (Up to the engines).
* It doesn't track changes. (Up to engines and stores).
* It doesn't make HTTP calls directly, for auth, downloading or uploading. (Up to the token and storage server clients).

mhammond replied:
Thanks - I added more about this and spoke specifically about telemetry.

thomcc commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+
+* Some apps only care about a subset of the engines - lockbox is one such app and only cares about a single collection/engine.
+
+* Some apps will use a combination of Rust components and "traditional" engines. For example, iOS is moving some of the engines to using Rust components, which other engines are likely to never be ported. We also plan to introduce some rust components into desktop in the same way, meaning desktop will eventually have both rust components and "traditional" engines involved.
```
Which engines on iOS aren't likely to be ported (Possibly worth calling out, just for completeness)

mhammond replied:
> Which engines on iOS aren't likely to be ported (Possibly worth calling out, just for completeness)

That's a great question for the long term - but over the next few months, I suspect we will consider things a complete success if the only engines which are ported are bookmarks and logins.

thomcc commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+
+There are a couple of gotchyas in the current implementations too - there's an issue when certain engines don't yet appear in meta/global - see bug 1479929 for all the details.
+
+The tl;dr of the above is that each rust component should be capable of working with different sync managers. In all cases, it is important that meta/global management is done in a holistic and consistent manner.
```
We also need to probably handle somehow the case that multiple sync managers are active at once.

This should still be doable during integration tests, however, so the constraint is probably something like 'one sync manager per engine'.

linacambridge replied:
I'd prefer we didn't have multiple sync managers at once in production. It makes a lot of sense for testing, but I think, just like with multiple `PlacesApi`s, multiple managers (with different or the same engines) would be a misuse of our API.

thomcc commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+
+# Approach for the rust components
+
+In general, we need to pass certain state around, between the "manager" and the engines. We must define this in a way so that the manager can be implemented anywhere - we should not assume the manager is implemented in rust.
```
Hmm, possibly this should be called out sooner. I definitely was assuming it would be.

Realistically I don't think we would be happy if something as big or complex as this were implemented outside of rust (It requires another sync15 implementation, for example), except possibly in the case of the current iOS implementation for now.

mhammond replied:
and possibly desktop? Longer term desktop could use rust, but I'd like to avoid a requirement to port the sync manager to desktop before we could experiment with replacing, say, the logins engine with a rust version.

thomcc commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+
+While we use rust struct definitions here, it's important to keep in mind that as mentioned above, we'll need to support the manager being written in something other than rust. As the data is relatively small, it may (or may not) be desirable to use JSON as the serialization format.
+
+    // The main structure passed to a sync implementation. This needs to carry
```
µnit: ` ```rust ` instead of the indent will give you rust syntax highlighting.

thomcc commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+        config: GlobalConfig,
+
+        // The token server token.
+        storage_token: Vec[u8] (??), // not clear on the type here, but whatever.
```
µnit: Well, probably needs to be `Vec<u8>` to be Rust syntax at least.

Would it be something like [`Arc<TokenserverToken>`](https://github.com/mozilla/application-services/blob/master/components/sync15/src/token.rs#L30) though? (Answer quite possibly could be no -- I don't know if this is the right token, and even if there is, maybe there's another reason it can't be that...)

mhammond replied:
I left this alone for now (although at face value, a clonable object seems a little better to me)

thomcc commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+
+        // The current syncID for the engine, as read from meta/global. May be None.
+        // Engines are responsible for persisting this and taking action when it differs.
+        sync_id: Option<SyncGuid>,
```
Version too, no? Or do we don't need to persist that since it can be a constant in the code?

mhammond replied:
This struct needs version too, yes. However, you are correct that it probably doesn't need to persist it, as it will be comparing it against a value that's hard-coded.

mhammond replied:
Added version and a note it may not need to persist.

thomcc commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+        commands: Vec<Command>
+    }
+
+     struct Command {
```
µnit: enum realistically, maybe.

mhammond replied:
Added a comment here noting it may be an enum

thomcc commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+## APIs
+## Sync Manager
+
+    // Register an engine.
```
µnit: ` ```rust `

linacambridge replied:
What if this were static, enabled by Cargo `features`? IOW, all available engines are in a static `HashMap`, gated at compile time...and the app passes a list of strings of the engines it wants to enable? (I _think_ that's what @thomcc suggested above, too, but I'm not sure).

So the set of supported engines is fixed, and, when the app does `SyncManager::new(&["bookmarks", "history"])`, or however that's translated over the FFI, this enables them.

mhammond replied:
I've expanded on this significantly - let me know if the new version looks better.

thomcc commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+## Sync Manager
+
+    // Register an engine.
+    fn register_engine(name: String, collections: Vec<String>, sync_callback: Fn<...>)
```
Also even if it's rust to rust, I think it would be really hard to make it work here due to unclear ownership. (It should be borrowed for this, probably, but who owns it then?)

mhammond replied:
Yeah, I should have chosen my words better here and avoided "callback" - I'll update the doc to something suitably vague ("a function to call" or something)

thomcc commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+# Details
+## Structure definitions
+
+While we use rust struct definitions here, it's important to keep in mind that as mentioned above, we'll need to support the manager being written in something other than rust. As the data is relatively small, it may (or may not) be desirable to use JSON as the serialization format.
```
JSON vs not JSON is basically a question of: Do we have to write the glue in e.g. Kotlin/Swift to pull data out of JSON. If so, then probably using protobufs are easier and cleaner, since that glue can be generated automatically.

mhammond replied:
I removed all refs to json.

linacambridge commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+
+* The sync manager is responsible for managing the "clients" collection - engines should never see details about that engine. However, engines will need the ability to be told about commands targetting that engine. The sync manager will also need to be able to accept commands to be delivered to other clients - although that needs more thought as the conditions in which a command need to be sent t ...
+
+* Each engine is responsible for persisting engine-specific stuff which needs to be persisted. An obvious thing which needs to be persisted is the syncID and version string for the engine itself, so it can take the necessary action when these change.
```
I'd prefer if the sync manager persisted the complete `m/g`. Each engine should also store the `syncID`, or some way to tie it to the current state, but we'll run the risk of uploading an incomplete `m/g` if we rely on engines to persist the data. (For example, what if we had an `m/g` entry for `addresses`, but we don't support the addresses engine?)

linacambridge replied:
Hmm, though...if we need to reupload `m/g`, say, because an engine's sync ID changes, we could always fetch it from the server, change _just that engine_ in the record, and upload, leaving the other engines alone.

mhammond replied:
I've expanded on this a little

linacambridge commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+
+In general, we need to pass certain state around, between the "manager" and the engines. We must define this in a way so that the manager can be implemented anywhere - we should not assume the manager is implemented in rust.
+
+We also need to consider the "clients" engine in this state - we will need to communicate commands sent by other devices and commands that we wish to send to other devices.
```
What if `clients` collection management were implemented as part of the manager, instead of as a separate engine? There's a fair bit of complexity for it in Desktop (can't be disabled, synced before and after, storage is in-memory without conflicts) that maybe it's not worth implementing as one—even if it uses similar abstractions for fetching and putting records.

mhammond replied:
Yes, clients should be handled by the manager - I've tried to clarify that.

linacambridge commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+
+* The sync manager will manage keys and pass the relevant key into the engine. Keys for engines should not be exposed to engines other than the key it is for (although currently, in practice, all engines share the same key, but this is an implementation detail.
+
+* The sync manager will pass the relevant part of info/collections to the engine and a token which can be used to access storage. Specific engines never talk to the token server.
```
Big :+1: on this and the above point!

linacambridge commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+    // it doesn't understand?
+    // Will return None if we haven't yet done a first sync so don't know. Will
+    // return an empty vec if we have synced but no engines are declined.
+    fn get_declined() -> Option<Vec<String>>
```
Ideally, this would all be internal to the sync manager, and we wouldn't expose any of this state to the app. (Except maybe via errors—for example, if Lockbox tries to sync, but you've declined passwords, we'll fail, and let them present UI for enabling).

I feel like global CWTS doesn't mesh well with single-purpose apps, where it's not obvious that un-declining on one un-declines on all, but that's a whole separate tangent.

mhammond replied:
I believe this will be needed so that the equivalent of about:preferences#sync can show what's been enabled or disabled?

grigoryk replied:
The consuming app needs to have a list of what's enabled/disabled for the UI. This could as well be `get_enabled`, to serve the same purpose - and it also might be a more straightforward API to actually use. The `disabled` bit is an internal detail of the sync1.5 protocol, after all.

linacambridge commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+    // in other circumstances, such as when wifi is lost but the user has
+    // opted in to syncing only on wifi
+    // TBD - does this make sense? If so, how long is the app expected to wait?
+    fn stop()
```
I think iOS inverts this a bit and has a "green light", where each engine periodically checks "can I keep syncing?" So I think calling `stop` would set a flag, and the engines would check that periodically as they do work.

mhammond replied:
Yes, that makes sense. Most engines would probably just reuse our existing `InterruptHandle`

mhammond commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+
+* The sync manager is responsible for persisting and populating the "declined" engine list. If an engine is declined, the manager should not ask it to sync. IOW, an engine doesn't really need to know whether it is enabled or not. However, engine should expose a way to be "reset" when the manager notices the declined state for an engine has been changed. This will allow engines to optimize certain ...
+
+* The sync manager is responsible for managing the "clients" collection - engines should never see details about that engine. However, engines will need the ability to be told about commands targetting that engine. The sync manager will also need to be able to accept commands to be delivered to other clients - although that needs more thought as the conditions in which a command need to be sent t ...
```
I suspect they will typically, but not always, come from the store - just not as a part of sync. For example, our store will be involved in a "restore bookmarks" operation, so could queue the command.

On the other hand, I can imagine a "delete all data from my account" operation, which might be implemented as something like "delete all local stores, then queue remote wipe commands for all other devices" - each store may not have enough context to queue those commands (as each store was just told to wipe locally and wipe the server, but doesn't know the broader operation being performed)

So to answer your question :) I was trying to say that we should not be too restrictive on where these commands can originate from as they *may* come from the app.

mhammond commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+    // Do a sync now. There's a special error code to indicate that something
+    // seems to have gone wrong with auth, which is surfaced to the app, so
+    // it can call on FxA to do magic.
+    fn sync_now(engines: Option<Vec<String>>) -> Result<()>
```
I suspect passing object references, which would include functions on the object, to be trickier than a simple function, but I'll see what I can do to make this more vague in this part of the doc and fleshing out a specific implementation plan later in the doc)

mhammond replied:
I made a few more notes about getting references to engines, and dropped all references to callbacks.

mhammond commented:
I've pushed a new revision which I hope addresses most of the comments above - see #5d6e33b for the diff, but the main changes are:

Main changes:

* Clarified that the tokenserver token may fail due to either auth issues or due to normal expiry (but for now I'm still sticking with the manager handling both those cases.

* Clarified some stuff around commands.

* Removed all references to callbacks, and made the (brave?) assertion that a rust implemented Sync manager will only ever deal with rust implemented engines, and further, that the engines and sync manager must all be in the same library.

* Removed all references to JSON

* Tried to clarify comments around the "not necessarily a 1:1 relationship between engines and collections"

* Added a few more things the manager will not do, and added some notes on telemetry.

* Clarified that the "clients" collection will be managed by the manager rather than as a new engine.

* Added the concept of a "prepare" function.

I'm going to resolve the conversations above which I hope I've suitably addressed - please open new conversations for what I've missed or made worse.

mhammond commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,171 @@
+    * The list of "declined" engines - that is, engines which the user has explicitly declined from syncing. Note that declined engines are global for the account - declining an engine on any device should decline it on every device. While this is something we'd like to consider changing in the future, for now we are sticking with this global declined state.
+    * Some metadata about every collection (which may be subtly different from an "engine"). Currently this metadata is only {version, syncID}
+
+    Management of meta/global includes all syncIDs - if there's no global syncID, or an engine has no syncID, it should generate one. As we mention below, engines will be responsible for persisting the syncID and version and taking necessary action when it changes. Thus, engines never need to communicate a new syncID back to the global.
```
I've tried to capture that as we discussed in slack, via a "prepare" function.

grigoryk commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,378 @@
+
+// Stop syncing ASAP. This might be called as the app is shutting down or when
+// the app has lost wifi, for example.
+fn stop();
```
I don't see something like Fenix using it in practice, as sync operations will be executed in background workers that survive app shutdowns.

grigoryk commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,378 @@
+// Do a sync now. There's a special error code to indicate that something
+// seems to have gone wrong with auth, which is surfaced to the app, so
+// it can call on FxA to do magic.
+fn sync_now(engines: Option<Vec<String>>) -> Result<()>;
```
`None` for engines means `sync all`?

grigoryk commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,378 @@
+// it doesn't understand?
+// Will return None if we haven't yet done a first sync so don't know. Will
+// return an empty vec if we have synced but no engines are declined.
+fn get_declined() -> Option<Vec<String>>;
```
Just to clarify expected usage... Given that list of disabled engines may change at any time:
- applications are expected to query this whenever they need to display a list of enabled/disabled engines
- to perform a `sync all`, applications should call `sync_now(None)`, instead of iterating list of what collections they think are enabled (this list could have changed)

Right?

mhammond replied:
My next version has changed this significantly - there's no separate API for enabling/disabling, or for determining what the current disabled states are. Instead, you specify state changes you desire before you sync, and at the end of the sync you get told what the current set is.

grigoryk commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,378 @@
+// it doesn't understand?
+// Will return None if we haven't yet done a first sync so don't know. Will
+// return an empty vec if we have synced but no engines are declined.
+fn get_declined() -> Option<Vec<String>>;
```
Also: I think `get_declined` and `sync_now` should somehow be symmetric, or it should be made obvious that they're not. Meaning that since collections and engines aren't 1:1, we should probably have a concept of "disabled engines" (or, conversely, "enabled engines"), instead of exposing an internal detail that is a "collection".

In other words - we shouldn't expose concept of collections at the API boundary at all.

rfk replied:
> In other words - we probably shouldn't expose concept of collections at the API boundary at all.

Strong +1, "collection" seems like a thing we map down to internally based on some higher-level concept.

mhammond replied:
Yeah, my new version does go into some detail about this.

* There's the concept of an "engine" - there's one called "history", but there are 2 "stores" behind this - "history" and "forms", and each of these stores currently uses exactly 1 collection.

* Imagine that in the future there was a "containers" collection so we can sync containers. In this case, the "history" store might actually use 2 collections.

So you could imagine a future where we have 1 "engine" which controls 2 "stores" which end up using 3 "collections".

As far as the sync manager is concerned, there are only "engines" - these are the things which are enabled or disabled.

grigoryk commented:
I left some inline comments.

Could you expand on how the proposed SyncManager will tie in with the individual storage layers? E.g. history has a "sync" method, and SyncManager has a sync(engines...) method. Will the rust syncmanager implementation invoke rust engines itself, passing in the state that it maintains?
It also wasn't entirely clear to me who's responsible for persisting that state. You mention that the state is FFI friendly to serve non-Rust implementations. I imagine in that case history/logins/etc storage library APIs will expose a "prepare" function over FFI, to set up the sync state, and the swift/js/whatever sync manager will be expected to call that as needed? And in the case of applications that just use the Rust implementation of the manager, they will not need to interact with prepare/sync functions exposed by the storage libs. Is that correct?

I'm assuming that Rust SyncManager will persist the state itself, since its APIs do not expose it.

mhammond commented:
I've pushed a new version which I hope addresses most of the previous comments. I'll re-flag everyone for review and setup a meeting to discuss this.

rfk commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,473 @@
+
+In practice, this means that the high-level component should only need to care
+about an *engine*. The low-level component will manage the mapping of engines
+to stores.
```
The app will independently need to know about stores though right, in order to actually use the store for providing application-level functionality?

Is it ever the case that a single store corresponds to multiple engines?  I expect yes, with the "places store" corresponding to both the "history" and "bookmarks" engines.

I wonder if there are any opportunities to clean up our historical baggage here.  For example, if we implemented future forms functionality as part of the places store, would that avoid the need to allow for multiple stores per engine?

mhammond replied:
> right, in order to actually use the store for providing application-level functionality?

Yes, this section is bad - naming is hard. It also gets to the root of the "sling instances around" comment above.

The best name, for the thing that provides application-level functionality, and given their names in the Rust code, is probably an "Api". The stores I refer to here are the [lower-level sync stores](https://github.com/mozilla/application-services/blob/master/components/sync15/src/sync.rs#L15). The "PlacesApi" has 2 sync-stores - bookmarks and history. The logins api has 1 sync store.

The sync manager can hopefully be done without knowing about any Apis and just about stores.

The framing of this part of the doc needs more work - it's from the inside looking out - thanks!

rfk commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,473 @@
+  This device does manage to write the new list to the server.
+
+* This device restarts and the network is up. It believes history is enabled
+  but logins is now - however, the state on the server is the exact opposite.
```
sp: "is not"?

rfk commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,473 @@
+component could avoid needing to persist any state, so it can avoid any
+kind of storage abstraction. A solution here might be that the low-level
+component can delegate state storage to the high-level component in an
+opaque way.
```
I will add that this is counter to the way many of our other components act, which is to handle their own storage transparently.  FxA is an outlier in this regard, but maybe we can follow some of its patterns around persisting stage?

rfk commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,473 @@
+// refer to what the user perceives as an "enigine" - but as above, these
+// *do not* correspond 1:1 with either "stores" or "collections" (eg, "history"
+// refers to 2 stores, and in a future world, might involve 3 collections).
+enum Engine {
```
We've inherited this name from the existing sync system, but this might also be a good opportunity to bikeshed a better one. I can't come up with a good rationale for why this is called an "engine", particularly in a world where most of the driving of a sync is done by the store.

rfk commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,473 @@
+  // thread.
+  fn interrupt() -> Result<()>;
+
+  // Disconnect this device from sync. This may "reset" the store, but will
```
nit: "stores" plural?

mhammond commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,473 @@
+
+* There may be some cross-crate issues even for the Rust implemented engines.
+  Or more specifically, we'd like to avoid assuming any particular linkage or
+  packaging of Rust implemented engines.
```
We'll usually need/want to ensure we have the right set of "instances" with references to each other - the existing singletons are all outside rust. This is certainly doable and not major though, so I'll tone that down in the doc.

mhammond commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,473 @@
+
+* Manage the "clients" collection - we probably can't ignore this any longer,
+  especially for bookmarks (as desktop will send a wipe command on bookmark
+  restore, and things will "be bad" if we don't see that command).
```
The client engine will want the FxA device ID, simply so it can populate `fxaDeviceId`. Desktop uses this for device names etc in various places.

It's probably ideal if we could use the fxa device ID *as* the client ID, which I'm leaning towards - in my mind, the only issue with that is that the FxA IDs are more ephemeral (ie, don't survive some password/account related events) - but is that actually a problem? We'd track this and clean up after ourselves (ie, we'd delete our old client record when we noticed the ID change, etc)

mhammond commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,473 @@
+* Define a minimal "service state" so certain things can be coordinated with
+  the high-level component. Examples of these states are "everything seems ok",
+  "the service requested we backoff for some period", "an authentication error
+  occurred", and possibly others.
```
I saw this as being completely transient from the component's POV - the API never needs it back. The embedding app might want to have it stick around in memory, but I don't think it ever needs to be persisted.

linacambridge replied:
I went back and forth on this after our meeting yesterday, but I like this approach. Keeping the service state coarse still lets the high-level sync manager recover, but avoids pushing complex error handling logic into the app. And, considering that the app will probably want to remember this state, anyway, it makes sense that we don't try to store it, too.

mhammond commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,473 @@
+## Clients Engine
+
+The clients engine includes some meta-data about each client. We've decided
+we can't replace the clients engine with the FxA device record and we can't
```
"Because the ID to too ephemeral"? :) As above, I'm rethinking that - it's a real PITA for client code to deal with them being different, so we should suck it up and see what breaks :)

linacambridge commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,473 @@
+
+* Enforce constraints necessary to ensure the entire ecosystem is not
+  subject to undue load. For example, this component should ignore attempts to
+  sync continiously, or to sync when the services have requested backoff.
```
Nit: continuously

linacambridge commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,473 @@
+  app's POV, there's a single "sync now" entry-point (in practice there might
+  be a couple, but conceptually there's a single way to sync). Note that as
+  mentioned above, this may involve coordinating with engines which aren't
+  implemented in Rust.
```
Does it make sense to push this into the high-level manager? We talked a bit about this yesterday, but it seems like trying to expose Rust thunks for existing Swift or JS engines is going to be tricky. Since Firefox for iOS and Desktop already have something like a sync manager, what if we had the high-level manager take care of the existing engines, and also call into the low-level Rust manager for Rust engines?

mhammond replied:
> Does it make sense to push this into the high-level manager?

Yeah, I agree it possibly does (although the low-level component will probably still need some knowledge for info/collections etc). I think we can see what shape we end up with before deciding this, so I added a new section "stuff tbd" and added this there.

linacambridge commented on docs/rfcs/sync-manager.md:
```diff
@@ -0,0 +1,473 @@
+different state request on a different device? Is the state as last-known on
+this device considered canonical?
+
+To clarify, consider:
```
_whispers_ Three-way merging. I'm joking a bit, but also...what if the sync manager persisted the list of declined engines, recorded local changes since the last sync, and merged them with the server's view once the network is back up? That does push some complexity for noting declined state changes up into the high-level manager, which could be an argument for having the sync manager manage its own storage.

As you say, though, it's an extreme edge case, and having the manager manage its own database is probably not worth the benefit of handling conflicting `declined` changes.

mhammond replied:
> _whispers_ Three-way merging.

We *could* do that - but *should* we?

Storing the transition is simple because it is a bool state - so just storing `(name, new_state)` seems enough - `old_state` can be implied. And we already store declined state, so adding this probably doesn't blow out "state budget" and force us into our own state storage.

It's more about user intent. It's like logins, where we *could* reconcile "device-a changed the username but not the password" and "device-b changed the password but not the username". We could reconcile and end up with both "new" values, but it seems unlikely to reflect user intent.

Similarly here - it's not clear to me "device-a declined bookmarks but left history enabled" but "device-b declined history and left bookmarks enabled" is best served by merging these rather than just taking the latest.

But yeah - "not clear" is the keyword here, so I'm erring on the side of simplicity - both in implementation and explanation - but I'm totally open to the fact I'm underthinking this :)
